packages/bsky/src/api/blob-resolver.ts

```diff
@@ -0,0 +1,79 @@
+import { pipeline, Readable } from 'stream'
+import express from 'express'
+import createError from 'http-errors'
+import axios, { AxiosError } from 'axios'
+import { CID } from 'multiformats/cid'
+import { ensureValidDid } from '@atproto/identifier'
+import { VerifyCidTransform } from '@atproto/common'
+import { NoResolveDidError } from '@atproto/did-resolver'
+import AppContext from '../context'
+import { httpLogger as log } from '../logger'
+
+// Resolve and verify blob from its origin host
+
+export const createRouter = (ctx: AppContext): express.Router => {
+  const router = express.Router()
+
+  router.get('/blob/:did/:cid', async function (req, res, next) {
+    try {
+      const { did, cid: cidStr } = req.params
+      try {
+        ensureValidDid(did)
+      } catch (err) {
+        return next(createError(400, 'Invalid did'))
+      }
+      let cid: CID
+      try {
+        cid = CID.parse(cidStr)
+      } catch (err) {
+        return next(createError(400, 'Invalid cid'))
+      }
+
+      const { pds } = await ctx.didResolver.resolveAtpData(did) // @TODO cache did info
+      const getBlob = await axios.get(`${pds}/xrpc/com.atproto.sync.getBlob`, {
+        params: { did, cid: cidStr },
+        decompress: true,
+        responseType: 'stream',
+        timeout: 2000, // 2sec of inactivity on the connection
+      })
+
+      const imageStream: Readable = getBlob.data
+      const verifyCid = new VerifyCidTransform(cid)
+
+      // Send chunked response, destroying stream early (before
+      // closing chunk) if the bytes don't match the expected cid.
+      res.statusCode = 200
+      res.setHeader(
+        'content-type',
+        getBlob.headers['content-type'] || 'application/octet-stream',
+      )
+      pipeline([imageStream, verifyCid, res], ignore)
```

Comment:
very slick

Reply:
altho do we want to handle errors at all? even just logging them or something? or will they get caught elsewhere?

Reply:
Hmmm good call!  I think they go into express and http-land and get handled there, since it's an error during response transmission which is a little atypical.  I will add a special log for it in here since it's pretty straightforward to do that.